Arrival Page Fixes

saintwolf opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
Fixes #1412
Fixes #1436
Fixes #2222

## Summary of Changes
* Fixed duplicate STARs showing for some Airports. In some cases (Such as EHAM), selecting ILS 08 would cause duplicate STARS on the arrivals list. This would also happen when selecting ILS 18. This fix checks to ensure the runway is exact (and also adds a check for Center runways if they exist at the airport).

* Fixed #1412. The last STAR on the list will now be scrollable. This also fixes #1436

* Approach Via page will now stop scrolling at the correct place. Previously the list would keep scrolling down, ad infinitum. 

* Arrival and departure pages will now display the selected Arrivals, Departures, Transitions, etc in yellow if there is a temporary flight plan, otherwise they will display in green.

* Approaches are now ordered as they should be on a Honeywell FMS.

* Arrival and departure pages both show runway course and length.

* Departure page runway list will stop scrolling in the correct place.


## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
### **Before:**
![image](https://user-images.githubusercontent.com/353949/101294535-c83c6c00-380f-11eb-8d28-8f375974d1c6.png)

![image](https://user-images.githubusercontent.com/353949/101294496-8d3a3880-380f-11eb-8c82-56647e294d4e.png)
### **After:**
![image](https://user-images.githubusercontent.com/353949/101846347-a52ff600-3b48-11eb-99c1-ad5efe804dbd.png)

![image](https://user-images.githubusercontent.com/353949/101846443-d8728500-3b48-11eb-9e46-3059581f1890.png)

### **Runway selection list changes:**
![image](https://user-images.githubusercontent.com/353949/101846602-3bfcb280-3b49-11eb-904c-9d51f6f66e7b.png)

![image](https://user-images.githubusercontent.com/353949/101847185-8d597180-3b4a-11eb-9b94-03bffc6a1c47.png)


## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

![image](https://user-images.githubusercontent.com/353949/101846151-42d6f580-3b48-11eb-9e23-238b60781439.png)


## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): Saintwolf#5557

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

Benjozork reviewed (approved):
LGTM

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:
```diff
@@ -54,7 +54,8 @@ class CDUAvailableArrivalsPage {
                                 if (runwayTransition) {
-                                    if (runwayTransition.name.indexOf(selectedRunway) != -1) {
+                                    // Check if selectedRunway matches a transition on the approach (and also checks for Center runways)
+                                    if (runwayTransition.name.match("^RW" + selectedRunway + "(C)*$")) {
```
```suggestion
                                    if (runwayTransition.name.match("^RW" + selectedRunway + "C?$")) {
```

Can be simplified to this (unless you want to match `RW09CCCCC`).

saintwolf replied:
Yeah this is much better. Tested it through a regex tester and looks like it works just as well :)

Highflyer2010 commented:
Category: QA Tester
Name: Discord User Highflyer2010 #8932
Date of testing: Current date 07/12/2020
Version of the sim: Version the tester uses 1.11.6.0
PR Tested: #2263 
Object of test: Checking Douplicate Stars showing at some airports (Without Navigraph)

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:

1. spawned at EGNT 
2. setting up a flight plan to EHAM
3. checking for duplicate entries ILS 08
4. No entry for ILS 08
5. Selecting ILS 18 - no duplicate entries

Flight notes (If required):

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: Make a small descriptive to areas where it needs to be improved and recommendations.

saintwolf commented:
@wpine215 Have added a Changelog

beheh commented:
Not required, but if you're touching the page anyway and still want to do this, one small change you could do is to update the button labels from the Thales style `TMPY ERASE` / `TMPY INSERT` to the Honeywell-style `{INSERT` / `ERASE*`. Here's a reference:

![image](https://user-images.githubusercontent.com/929769/101610104-14ea9780-3a08-11eb-84cb-3d0160536d0e.png)

saintwolf commented:
@beheh No problem! It looks like the Thales style is used in a few files: 
- A320_Neo_CDU_AvailableArrivalsPage.js
- A320_Neo_CDU_AvailableDeparturesPage.js
- A320_Neo_CDU_DirectToPage.js
- A320_Neo_CDU_FlightPlanPage.js

Should probably update them all for consistency. Should I do them all here or create a new PR?

beheh commented:
I think some of it is just still from Asobo, and we weren't Honeywell from day 1 either. The Flight Plan page is already being worked on in #2245, but the others would be great, thanks!

saintwolf commented:
@beheh All done!

![image](https://user-images.githubusercontent.com/353949/101673906-b0ebc180-3a4f-11eb-8056-b84c3d82bfec.png)
![image](https://user-images.githubusercontent.com/353949/101673955-c2cd6480-3a4f-11eb-8807-7763e49361a0.png)
![image](https://user-images.githubusercontent.com/353949/101673994-d37dda80-3a4f-11eb-9da8-9d9bc76f5df6.png)

saintwolf commented:
Made some more changes to align the departures page with the Honeywell design: 
![image](https://user-images.githubusercontent.com/353949/101677629-b5ff3f80-3a54-11eb-8f38-2a9f254eeaea.png)

Reference: https://www.theairlinepilots.com/forumarchive/a320/a320-fms-pages.pdf

beheh commented:
If you want to go the extra mile for accuracy, add a few non-breaking spaces before some of the labels:

![image](https://user-images.githubusercontent.com/929769/101688582-86f4c800-3a6c-11eb-9581-1195f96553a4.png)

They're a bit hard to spot at first, but you can't un-see them once you realize they're there :-)
You can use `{sp}` [like I did here](https://github.com/flybywiresim/a32nx/pull/2262/files).

saintwolf commented:
Added spaces, and updated some other text as per the FMS reference:

![image](https://user-images.githubusercontent.com/353949/101790735-b5b97f80-3afa-11eb-99f9-2cb3017b3dcc.png)
![image](https://user-images.githubusercontent.com/353949/101790869-d4b81180-3afa-11eb-8f4a-7090dbfe924e.png)

beheh commented:
Two more things: From my last reference, the selected items are yellow while in a tempy flight plan (when INSERT/DELETE is shown) and the header could be updated to say `ARRIVAL {small}TO{end} {green}EDDF{end}`

saintwolf commented:
@beheh How's this look?

![image](https://user-images.githubusercontent.com/353949/101803801-d12b8700-3b08-11eb-92f3-c183ea4f1739.png)
![image](https://user-images.githubusercontent.com/353949/101803886-eb656500-3b08-11eb-8149-05814356ab46.png)
![image](https://user-images.githubusercontent.com/353949/101803958-fe783500-3b08-11eb-9cf0-dad607a23eec.png)
![image](https://user-images.githubusercontent.com/353949/101804064-15b72280-3b09-11eb-8784-3f3105965fcd.png)
![image](https://user-images.githubusercontent.com/353949/101804141-29628900-3b09-11eb-8aeb-28b6e3b264b8.png)

saintwolf commented:
Okay I think I'm done with all of the changes. @beheh @Benjozork Can you take a look at the updated code please?

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:
```diff
@@ -5,38 +5,68 @@ class CDUAvailableArrivalsPage {
             const approaches = airportInfo.approaches;
+            // Sort the approaches in Honeywell's documented order
+            const sortedApproaches = approaches.sort(
+                function (a, b) {
+                    const approachTypeOrder = {"MLS":0, "ILS":1, "GLS":2, "IGS":3, "LOC":4, "BLOC":5, "LDA":6, "SDF": 7, "GPS": 8, "RNAV":9, "VORDME":10, "NDB":11};
+                    return approachTypeOrder[a.name.split(" ")[0]] - approachTypeOrder[b.name.split(" ")[0]];
+                }
```
Use an arrow function instead.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:
```diff
@@ -5,38 +5,68 @@ class CDUAvailableArrivalsPage {
+                        const approachRunwayName = Avionics.Utils.formatRunway(approach.name.split(" ")[1]);
+                        let runwayLength = 0;
+                        let runwayCourse = 0;
+                        for (let i = 0; i < runways.length; i++) {
+                            const runway = runways[i];
```
Use a `for ... of` loop instead.

BritishAvgeek commented:
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: Current date 14.12.2020
Version of the sim: 1.11.6.0
PR Tested: #2263 
Object of test: Arrival Page Fixes

Testing technique: Full Flight livestreamed here https://youtu.be/tcRMJ7YzaS0    Mumbai to Delhi
Steps you've made to test the Pull Request:
1. Generated flight plan using SimBrief linked to Navigraph and AIRAC 2013
2. Spawned C&D - Input plan into MCDU as per plan
3. During input of departure, it reverted to it's behaviour pre-fix so the Temp FPLAN and yellow for the departure no longer showed
4. During input of Planned Arrival, it was all off - I had to select ILS 27 for Runway 28 - but it then gave me VOR DME 28 instead of ILS - but still input the ILS freq into the RADNAV (see comments below)
5. Flew route and landed, no issues through flight except the wider known sim issues

Flight notes (If required):
Route: VABB/09 ALBA2B ISRIS DCT DOSTO DCT EGUGU DCT GOMTI DCT INTIX DCT NOKOK DCT BUBNU DCT DUKOT DCT ELMON ATRI5B/28

Flew the VOR DME arc manually anyway on arrival. It wasn't incorrect in forcing the VOR DME 28 but bearing in mind the selection was ILS 27 to trigger the button below which read ILS 28 - it should have displayed ILS 28 instead and fed the RNAV arrival for ATRI5B in, instead of the DME ARC. I flew the ARC anyway for practice and to show people watching the stream how to fly it but the input into the MCDU seemed wrong

Go to:
https://www.youtube.com/watch?v=tcRMJ7YzaS0&feature=youtu.be
Relevant Times:
06mins 10 seconds for Departure Input to show the issues now arising from that - gave us just "ERASE" or "INSERT" as options for the SID - not a return to flight plan option 
10mins 40seconds onwards for Arrival Input issues but I selected ILS28 and it gave me VORDME29 - I selected ILS27 and it gave me VORDME28
2:00:00 onwards for the VOR DME Arrival and ILS approach

Issues (If Any): 
Overall Rating: To be improved and needs further testing
Conclusions of the testing: All mentioned above - unsure as to whether the VORDME read properly on the arrival page but the input was for an ILS approach (which is fed into the MCDU RADNAV page anyway) not a VOR DME which I believe should have been input as the navigraph screenshot attached shows:
![image](https://user-images.githubusercontent.com/71381591/102106526-98512200-3e28-11eb-807e-55a9b879d63d.png)


I ended up flying the DME ARC anyway as shown below as this is also what it was forcing (orange line):
![image](https://user-images.githubusercontent.com/71381591/102106629-b61e8700-3e28-11eb-81b9-11b2d8423751.png)

beheh commented:
> 06mins 10 seconds for Departure Input to show the issues now arising from that - gave us just "ERASE" or "INSERT" as options for the SID - not a return to flight plan option

That behavior sounds correct - there should only be ERASE and INSERT, and you can click FPLN if you want to see the flight plan. Pressing INSERT should insert and take you to the flight plan immediately.

![image](https://user-images.githubusercontent.com/929769/102114064-1669f480-3e3a-11eb-9223-0faebaf30f1c.png)

BritishAvgeek commented:
What about the rest?
That wasn't as much an issue but an observation of it working differently but the actual arrival runway selection was seemed all wrong with the wrong runways and approaches being chosen from input etc.

beheh commented:
> Also should the departure selection and pressing Insert not keep it as a temp flight plan as it did before then returning to the FLT PLN page displaying all as yellow/temp? This one committed the SID in green and I had to go in and trigger it into temp Flt PLN by continuing input of the route.

I thought it behaved exactly like this before, but in any case, no: "Insert" immediately overwrites the current, active flight plan. If you want to see a preview you need to explicitly press the FPLN button.

![image](https://user-images.githubusercontent.com/929769/102121155-ca23b200-3e43-11eb-95fa-8e3fdc7e7714.png)

BritishAvgeek commented:
Cheers for that, good to know... Still the arrivals page issues need looking into, as they appeared quite wrong for my destination airport today

SierraBravo1 commented:
The mismatch with the approach selection on the Arrival page is also present for an arrival to EDDN.
The approach selection is shifted by one item.

If I select -> I get:

ILS 10  -> VORDME 10
ILS 28Z -> ILS 10
LOC 28Y -> ILS 28
RNAV 10 -> LOC 28
VORDME 10 -> RNAV 10

saintwolf commented:
Thanks for the testing folks, I know why this is happening now. Basically I am reordering the array of approaches, however the index of that approach is required to select it afterwards. Does anyone know if this index is stored within the approach object at all? E.g. with something like getApproachIndex()?

SierraBravo1 commented:
hanks, saintwolf.
I found another issue with the arrival page. It would be great if you could have a look at it.

Set EDDS as destination.
Select approach ILS 07
Select STAR DENE1E
Select No VIA

DENE1E arrival is added, but the approach via BADSO is inserted automatically, which cannot be removed.
Both, STAR DENE1E and the approach via BADSO are active in the flightplan.
It looks like it is not possible to select NO VIAS. If done, the BADSO approach is included in addition.

![EDDS_DENE1E](https://user-images.githubusercontent.com/76132388/102510461-f29bee00-4087-11eb-9166-5f4a8f956abf.png)

saintwolf commented:
@SierraBravo1 There is no special handling in the menu for the No VIAs option (it's built in), it is likely something to do with Asobo's flight plan manager, which I am unable to fix in this code unfortunately.

saintwolf commented:
@SierraBravo1 I think this is due to the function call setApproachTransitionIndex(-1) which is not being handled correctly, however it's not something I can fix in the scope of the arrival page code unfortunately. This would need a separate issue/PR to investigate (if we can even edit the flight plan manager code without affecting other aircraft).

Awemeter commented:
Category: QA Tester
Name: Awemeter#7919
Date of testing: 12/17/2020
Version of the sim: 1.11.7.0
PR Tested: #2263 
Object of test: Arrival pages. 

Testing technique: Full flight on a route that I know has double missing stars.
Steps you've made to test the Pull Request: 
1. C&D Full flight Kord to Kdca. 
2. Programmed flight plan and did not see double stars. I also saw the enhancements on the arrival page.
3. Landed at DCA


Flight notes (If required): 

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: A very great pr that allows me to fly into airports with a messed up arrival page. Thanks for your hard work and keep it up.

Benjozork commented:
> @SierraBravo1 I think this is due to the function call setApproachTransitionIndex(-1) which is not being handled correctly, however it's not something I can fix in the scope of the arrival page code unfortunately. This would need a separate issue/PR to investigate (if we can even edit the flight plan manager code without affecting other aircraft).

We can, just gotta copy it over. But that's already being worked on by @wpine215. There's a proposal in flybywiresim/rfcs

SierraBravo1 commented:
> 
> 
> Thanks for the testing folks, I know why this is happening now. Basically I am reordering the array of approaches, however the index of that approach is required to select it afterwards. Does anyone know if this index is stored within the approach object at all? E.g. with something like getApproachIndex()?

Fixed for me
Tested approach selection of EDDN and LFMN successfully.
Thanks

tjburni commented:
Category: QA Tester
Name: [Z+1]tjburni#0409
Date of testing: Current date 19.12.2020
Version of the sim: 1.11.7
PR Tested: #2263
Object of test:Arrival Pages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Full IFR Flight from C&D 
2. Programmed flight plan and checked for the fixed issues descrobed in the pr
3. Landed
4.
5.

Flight notes (If required):

Issues (If Any): none
Overall Rating: Good
Conclusions of the testing: Looks good for me all issues are gone. Thanks
